### PR TITLE
Simulation code for llm inference gateway

### PR DESCRIPTION
**Pull request by kaushikmitr**
## The notebook implements the following components:

1. **Prefill and Decode Latency Modeling as a Function of Batch Sizes**:
   - The prefill and decode latencies are modeled based on batch size, where the total latency is influenced by the number of tokens in a batch.
   - The model incorporates different latency constants, providing a mathematical representation of how batch sizes impact the prefill and decode processes.
   - Latency is calculated using a formula that sums up token lengths across the batch, giving insight into the relationship between batch size and latency performance.

2. **Continuous Batching Algorithm Simulation**:
   - This component simulates a continuous batching mechanism where scheduling decisions are made dynamically before each forward pass as implemented in vLLM.

3. **LLM Inference Gateway Simulation with Different Routing Strategies**:
   - The simulation models an LLM inference gateway, where requests are routed to different model servers based on various strategies.
   - Multiple routing algorithms are explored to optimize how requests are handled, balancing latency, throughput, and resource utilization.

4. **Evaluating Different Algorithms with Configurable Workloads**:
   - The notebook enables the evaluation of various algorithms under different configurable workloads, allowing for flexible testing of different conditions.
   - Users can configure workloads by modifying factors such as request sizes, batch sizes, token lengths, and latency requirements.

### CODE REVIEW

**Comment by ahg-g** [comment]
any idea to make this reviewable on github? the file is way too big with image blobs taking up so much space

**Comment by kaushikmitr** [comment]
> any idea to make this reviewable on github? the file is way too big with image blobs taking up so much space

Yes, this is more for viewing (https://github.com/kubernetes-sigs/llm-instance-gateway/blob/a6ee825c4a91d4a5c7915d82b6a6991123347c7b/simulations/llm_inference_gateway_simulation.ipynb) but hard to comment on. I am working on making this into a regular python script.

**Comment by kaushikmitr** [comment]
added python library llm_ig_simulation folder

File: simulations/llm_ig_simulation/src/benchmark_one_server.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
+
+        
+        
+
```
Comment (terrytangyuan) [inline simulations/llm_ig_simulation/src/benchmark_one_server.py:4]: Remove extra lines (here and other places)

Comment (kaushikmitr) [reply simulations/llm_ig_simulation/src/benchmark_one_server.py]: done

**Comment by ahg-g** [comment]
When you get a chance, can we have readme file describing how to run this?

**Pull request merged**